multimodal search pdf

Yongxuanzhang opened the pull request:
WIP

Design:
Index FLow:


Query Flow with 3 segmenters:
This design will cause flaky issues.
![image](https://user-images.githubusercontent.com/44033547/112388813-1b0c3780-8cca-11eb-9d85-5ba2f3473021.png)
Use this compound executor could work
![image](https://user-images.githubusercontent.com/44033547/112573756-4e7ebd00-8dc3-11eb-9340-55433d2dc658.png)



- [x] Search image,text and pdf in one flow or separate flows
- [x] Should I divide the multimodal segmenter(in query flow) into 3 separate segmenters?
- [x] Having 1 chunk for all text first, then segment it into chunks by paragraph or lines?
-  Should chunk the text into smaller chunks(like lines or sentences instead of large chunks)
- [x] How to view the pdf results(also test rest API)?
- Jinabox needs frontend supports
- Right now I just print URI and score of the matches, this can help us verify if the query is correct
- [x] Other pdfs(now test on papers from arxiv)->should have a good datatset
- Note that text of pdf download from Jina blog cannot be extracted of PyPDF2
- [x] Probably cannot deal with 1 channel and 3 channels images at the same time (for 1 pdf file)
- change the pdf segmenter, to craft the images with 1 channel and 4 channels into 3channels array
- [x] should achieve high search accuracy
- [x] add parallel peas to decrease overall indexing time
- [x] add docker file and manifest
- [x] add test 
- [x] add readme
- [x] add docstrings for all YAML files
- [x] use better pdf extractor with @Kelton8Z 's work [ticket](https://github.com/jina-ai/internal-tasks/issues/38)
- [x] add performance
- [ ] Other features(optional)?
- Add page number of text and images
- Leverage more CompoundExecutors
- Frontend support

Yongxuanzhang committed:
feat[mmpdf]: pdf search draft

Yongxuanzhang committed:
feat[mmpdf]: code refactoring

theUnkownName commented on multimodal-search-pdf/pods/PDFcrafter.py:
```diff
@@ -0,0 +1,183 @@
+from jina.executors.segmenters import BaseSegmenter
+
+
+class PDFSegmenter(BaseSegmenter):
```
I didn't get yhy are we copying the PDFSegmenter here? And why we call it crafter?

Yongxuanzhang replied:
Because I want to modify it. Like for text, I want to segment it into lines here.
Thanks for reminding me of that. I thought I have changed the name 😅

FionnD commented:
> * [ ]  How to view the pdf results(also test rest API)?
In theory we could use JinaBox, if you wanted we could talk to FE about building something nicer. But that needs conversation needs to go via Fionn/Agnes. 

> * [ ]  Other pdfs(now test on blogs from Jina website)
What dataset are we using atm?

bwanglzu commented on multimodal-search-pdf/app.py:
```diff
@@ -0,0 +1,72 @@
+
+import click
+from jina import Document
+from jina import Flow
```
don't forget to clean up the import

bwanglzu commented:
please keep an eye on @Kelton8Z 's [ticket](https://github.com/jina-ai/internal-tasks/issues/38), seems pdfminer is more robust compared with our current pdfsegmenter.

rutujasurve94 commented on multimodal-search-pdf/pods/PDFcrafter.py:
```diff
@@ -0,0 +1,183 @@
+
+        if uri:
+            pdf_img = fitz.open(uri)
+            pdf_text = open(uri, 'rb')
```
Open all files using the `with` context for better error handling

Yongxuanzhang replied:
Thanks! I will check this!

rutujasurve94 commented on multimodal-search-pdf/pods/PDFcrafter.py:
```diff
@@ -0,0 +1,183 @@
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def _pdf_segment(self, uri: str, buffer: bytes, *args, **kwargs) -> List[Dict]:
```
This is exactly the same as the `PDFSegmenter` class above, we could simply use an object of the class and the methods underneath?
Might be a better idea to have three separate classes - segment pdf, segment text and segment images, all inherited from a common superclass

Yongxuanzhang replied:
Yes, that's one point I put in the description of this PR, I also prefer to separate it into three, so I can add drivers for each. I will try it later

rutujasurve94 commented on multimodal-search-pdf/pods/PDFcrafter.py:
```diff
@@ -0,0 +1,183 @@
+
+        chunks = []
+        # Extract images
+        with pdf_img:
```
Add `docstring` briefly describing the reshaping, pixmap operations and their use cases

rutujasurve94 commented on multimodal-search-pdf/pods/craft-mminput.yml:
```diff
@@ -0,0 +1,16 @@
+jtype: CompoundExecutor
```
Rename file to a better name

Yongxuanzhang replied:
Good catch! This is one I wanted to modify but totally forgot. 😂

Yongxuanzhang commented:
> > * [ ]  How to view the pdf results(also test rest API)?
> 
> 
> 
> In theory we could use JinaBox, if you wanted we could talk to FE about building something nicer. But that needs conversation needs to go via Fionn/Agnes. 
> 
> 
> 
> > * [ ]  Other pdfs(now test on blogs from Jina website)
> 
> 
> 
> What dataset are we using atm? 
> 
> 
> 
 Yes that is what I want. Current Jinabox doesn't support pdf data. I wonder how much engineering effort we need to add that? 
The data I'm using are blogs download from our website in Chrome

rutujasurve94 commented on multimodal-search-pdf/pods/crafte-image.yml:
```diff
@@ -0,0 +1,13 @@
+jtype: ImageNormalizer
```
rename file to `craft-image.yml` instead of `crafte-image.yml`

rutujasurve94 commented on multimodal-search-pdf/pods/index-comp.yml:
```diff
@@ -0,0 +1,50 @@
+requests:
+  use_default: true
+  on:
+    SearchRequest:
```
Add docstring describing why we use `c` path for Vector and `cm` for KV

Yongxuanzhang commented:
> please keep an eye on @Kelton8Z 's [ticket](https://github.com/jina-ai/internal-tasks/issues/38), seems pdfminer is more robust compared with our current pdfsegmenter.

Sure! Thanks! That's something in my plan! And That's also the reason I used a customized segmenter atm.

rutujasurve94 commented on multimodal-search-pdf/pods/ranker.yml:
```diff
@@ -0,0 +1,24 @@
+  use_default: true
+  on:
+    SearchRequest:
+      - jtype: Chunk2DocRankDriver
```
Add docstring for describing `Chunk2DocRankDriver` usage, and why we use `sortQL` and `sliceQL` filters

rutujasurve94 commented on multimodal-search-pdf/pods/PDFcrafter.py:
```diff
@@ -0,0 +1,183 @@
+        """
+
+
+        import fitz
```
Briefly add the use of `fitz` library in the docstring

rutujasurve94 commented:
Thanks for this great example on multimodal using `PDFExtractor`, it covers three complex parallel workflows of pdf, text and image. Great work @Yongxuanzhang!

Added few comments for the example, add the `weighted_ranker.py` module as well.

Yongxuanzhang commented on multimodal-search-pdf/pods/PDFcrafter.py:
```diff
@@ -0,0 +1,183 @@
+        if text:
+            return self._text_segment(text)
+        elif uri:
+            from jina.types.document.converters import guess_mime
```
I remembered this is done in `document`, will check and fix this if so

Yongxuanzhang commented:
> Thanks for this great example on multimodal using `PDFExtractor`, it covers three complex parallel workflows of pdf, text and image. Great work @Yongxuanzhang!
> 
> Added few comments for the example, add the `weighted_ranker.py` module as well.

Thanks for all your insightful comments, really help a lot!! 😄

Yongxuanzhang committed:
feat[mmpdf]: fix text segment

Yongxuanzhang committed:
feat[mmpdf]: fix mimetype

Yongxuanzhang committed:
feat[mmpdf]: divide segmenters

Yongxuanzhang committed:
feat[mmpdf]: new pdf extractor

Yongxuanzhang committed:
feat[mmpdf]: parallel segmenter

Yongxuanzhang committed:
feat[mmpdf]: deeper chunks

maximilianwerk committed:
fix: split indexer into c and cc

Yongxuanzhang committed:
feat[mmpdf]: separate text query

Yongxuanzhang committed:
feat[mmpdf]: fix mminput

Yongxuanzhang committed:
feat[mmpdf]: fix ranker

Yongxuanzhang committed:
feat[mmpdf]: clean code

Yongxuanzhang committed:
feat[mmpdf]: clean code

Yongxuanzhang committed:
feat[mmpdf]: add docstrings

Yongxuanzhang committed:
feat[mmpdf]: add docstrings

Yongxuanzhang committed:
feat[mmpdf]: add docstrings

Yongxuanzhang committed:
feat[mmpdf]: add docstrings

Yongxuanzhang committed:
feat[mmpdf]: clean code

rutujasurve94 commented on multimodal-search-pdf/pods/index-comp-text-cc.yml:
```diff
@@ -0,0 +1,50 @@
+requests:
+  use_default: true
+  on:
+    SearchRequest:
```
Add docstring for traversal path describing why we use `cc`, `ccm`, describe why we exclude `embedding` and `blob`  in `ExcludeQL`

Yongxuanzhang replied:
Thanks! @rutujasurve94, I just addressed these comments 😄

rutujasurve94 commented on multimodal-search-pdf/requirements.txt:
```diff
@@ -0,0 +1,9 @@
+jina[scipy, http]==1.0.15
+PyPDF2==1.26.0
+PyMuPDF==1.18.5
+Pillow==8.0.1
```
Use `Pillow==8.1.2`, since `8.0.1` has security issues

rutujasurve94 commented on multimodal-search-pdf/pods/segmenter-pdf.yml:
```diff
@@ -0,0 +1,12 @@
+metas:
+  name: seg-pdf
+  py_modules:
+    - PDFSegmenter-separate.py
```
Rename file, use camelcase convention only

rutujasurve94 commented on multimodal-search-pdf/pods/segmenter-pdf.yml:
```diff
@@ -0,0 +1,12 @@
+jtype: PDFSegmenter
+metas:
+  name: seg-pdf
```
Rename to segment_pdf

Yongxuanzhang committed:
feat[mmpdf]: fix none text

JoanFM commented on multimodal-search-pdf/flows/index.yml:
```diff
@@ -0,0 +1,42 @@
+jtype: Flow                                 # this YAML file define the index flow, jtype is the YAML tag
```
Being such a large example? should we aim to merge some Executors into CompoundExecutors to have less processes? It may be much more optimal

Yongxuanzhang replied:
Yes, that makes sense! I will try to add a compressed flow

Yongxuanzhang committed:
feat[mmpdf]: fix channels

Yongxuanzhang committed:
feat[mmpdf]: add parallel

Yongxuanzhang committed:
feat[mmpdf]: address comments

Yongxuanzhang committed:
feat[mmpdf]: rename file

Yongxuanzhang committed:
feat[mmpdf]: add docstrings

Yongxuanzhang committed:
feat[mmpdf]: fix generator

Yongxuanzhang committed:
feat[mmpdf]: add index generator

Yongxuanzhang committed:
feat[mmpdf]: add docker file

Yongxuanzhang committed:
feat[mmpdf]: clean app

Yongxuanzhang committed:
feat[mmpdf]: add test

Yongxuanzhang committed:
feat[mmpdf]: add rest endpoint

Yongxuanzhang committed:
feat[mmpdf]: add manifest

Yongxuanzhang committed:
feat[mmpdf]: add ci

Yongxuanzhang committed:
feat[mmpdf]: add toy data

Yongxuanzhang committed:
feat[mmpdf]: add toy data

Yongxuanzhang committed:
feat[mmpdf]: remove pypdf2

Yongxuanzhang committed:
feat[mmpdf]: loose test

Yongxuanzhang committed:
feat[mmpdf]: pump version

Yongxuanzhang committed:
feat[mmpdf]: add more docstrings

Yongxuanzhang committed:
feat[mmpdf]: fix dryrun

Yongxuanzhang committed:
feat[mmpdf]: fix rest

Yongxuanzhang committed:
feat[mmpdf]: add doc

Yongxuanzhang committed:
feat[mmpdf]: add readme

rutujasurve94 commented:
@Yongxuanzhang if this is fully functional we should be good to merge this?
you could add metrics around index time, evaluation, etc. here to get an idea. would be great to have a summary of the latest functional changes for this example. Remove redundant / duplicate class files.

Yongxuanzhang commented:
> @Yongxuanzhang if this is fully functional we should be good to merge this?
> you could add metrics around index time, evaluation, etc. here to get an idea. would be great to have a summary of the latest functional changes for this example. Remove redundant / duplicate class files.

No problem! Where should I put the summary? I think readme?

JoanFM commented:
It may make sense to merge some of the pods with `CompoundExecutors` to avoid having so many services. It may speed up the processes and help people onboard to it by having less files

Yongxuanzhang commented:
> It may make sense to merge some of the pods with `CompoundExecutors` to avoid having so many services. It may speed up the processes and help people onboard to it by having less files

Thanks! I will merge them for sure

Yongxuanzhang committed:
feat[mmpdf]: fix jpeg

rutujasurve94 commented:
Yes, summary could be added to the README

Yongxuanzhang committed:
feat[mmpdf]: fix image seg

Yongxuanzhang committed:
feat[mmpdf]: add docstrings

JoanFM commented on multimodal-search-pdf/pods/segmenters.py:
```diff
@@ -164,7 +164,7 @@ def segment(self,  uri: str, buffer: bytes, mime_type: str, *args, **kwargs) ->
         chunks=[]
         if uri:
-            if mime_type == 'image/png' or 'image/jpeg':
+            if mime_type == 'image/png' or mime_type == 'image/jpeg':
```
if `mime_type` in [''mage/png' , 'image/jpeg']
```suggestion
            if `mime_type` in [''mage/png' , 'image/jpeg']:
```

Yongxuanzhang replied:
how about changing it to `if mime_type.startswith('image'):`?

JoanFM replied:
can be done as well yes

Yongxuanzhang committed:
feat[mmpdf]: fix segmenter

Yongxuanzhang committed:
feat[mmpdf]: fix image segmenter

nan-wang commented on multimodal-search-pdf/pods/segmenters.py:
```diff
@@ -0,0 +1,285 @@
+        :rtype: List[Dict]
+        """
+        chunks=[]
+        if mime_type == 'application/pdf':
+            import fitz  # fitz is a library used in `PyMuPDF` to read pdf and images
```
A long code block between the `if` clause and the `return` makes the codes difficult to read
```suggestion
        if mime_type != 'application/pdf':
            return chunks
        import fitz  # fitz is a library used in `PyMuPDF` to read pdf and images
```

nan-wang commented on multimodal-search-pdf/pods/segmenters.py:
```diff
@@ -0,0 +1,285 @@
+        :returns: A list of documents with the extracted data
+        :rtype: List[Dict]
+        """
+        chunks=[]
```
```suggestion
        chunks = []
```

nan-wang commented on multimodal-search-pdf/pods/segmenters.py:
```diff
@@ -0,0 +1,285 @@
+            else:
+                raise ValueError('No value found in "buffer" or "uri"')
+
+            chunks = []
```
duplicates with `L35`
```suggestion
```

Yongxuanzhang committed:
feat[mmpdf]: add readme

Yongxuanzhang committed:
feat[mmpdf]: add readme

nan-wang commented on multimodal-search-pdf/pods/segmenters.py:
```diff
@@ -0,0 +1,285 @@
+        return chunks
+
+
+class MultimodalSegmenter(BaseSegmenter):
```
It is better to inherit from the `PDFSegmenter` because they share quite a lot of codes

Yongxuanzhang replied:
How about removing this class? I intend to keep it just as a backup and not used in the current flow.

Yongxuanzhang replied:
removed

nan-wang commented on multimodal-search-pdf/pods/segmenters.py:
```diff
@@ -0,0 +1,285 @@
+
+class PDFSegmenter(BaseSegmenter):
+    """
+    :class:`PDFExtractorSegmenter` Extracts data (text and images) from PDF files.
```
```suggestion
    :class:`PDFSegmenter` Extracts data (text and images) from PDF files.
```

Yongxuanzhang committed:
feat[mmpdf]: add readme

nan-wang commented on multimodal-search-pdf/pods/segmenters.py:
```diff
@@ -0,0 +1,285 @@
+        return chunks
+
+
+class TextSegmenterCustomized(BaseSegmenter):
```
Make it clear that it is inherited from `Segmenter`
```suggestion
class CustomizedTextSegmenter(BaseSegmenter):
```

nan-wang commented on multimodal-search-pdf/pods/segmenters.py:
```diff
@@ -0,0 +1,285 @@
+                    chunks.append(dict(text=text_page, weight=1.0, mime_type='text/plain'))
+        return chunks
+
+    @single
```
`@single` has already been added to `segment()`
```suggestion
```

nan-wang commented on multimodal-search-pdf/pods/segmenters.py:
```diff
@@ -0,0 +1,285 @@
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @single(slice_nargs=3)
```
As it is `single`, there is no need to use `slice_nargs`
```suggestion
    @single
```

Yongxuanzhang replied:
remove this will not work I think. I tried to remove it but then no results returned.

Yongxuanzhang replied:
And this parameter should be set  to equal to the number of arguments if number of the arguments is larger than 1

nan-wang replied:
hmm, that is interesting. I will dig into this

JoanFM replied:
@Yongxuanzhang is right. Single speaks about instances passed, not about the amounts of arguments

nan-wang commented on multimodal-search-pdf/pods/segmenters.py:
```diff
@@ -0,0 +1,285 @@
+                                dict(blob=np_arr, weight=1.0, mime_type='image/png'))
+
+            # Extract text
+            with pdf_text:
+                count = len(pdf_text.pages)
+                for i in range(count):
+                    page = pdf_text.pages[i]
+                    text_page = page.extract_text(x_tolerance=1, y_tolerance=1)
+                    if text_page:
+                        chunks.append(dict(text=text_page, weight=1.0, mime_type='text/plain'))
```
Duplicates with `L230`
```suggestion
            with pdf_text:
                count = len(pdf_text.pages)
                for i in range(count):
                    page = pdf_text.pages[i]
                    text_page = page.extract_text(x_tolerance=1, y_tolerance=1)
                    if text_page:
                        chunks.append(dict(text=text_page, weight=1.0, mime_type='text/plain'))
```

Yongxuanzhang replied:
The code of L230 is actually another class (MultimodalSegmenter), and the class is not used in the current flow. 
It's a combination of 3 segment functions(image,text and pdf). It can be used if we want to have all segmenters in 1 class. I'm not sure if I need to keep it or not.

Yongxuanzhang replied:
Rmoved

nan-wang commented on multimodal-search-pdf/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+FROM jinaai/jina:1.0.15
```
```suggestion
FROM jinaai/jina:1.1.0
```

nan-wang commented on multimodal-search-pdf/README.md:
```diff
@@ -0,0 +1,129 @@
+
+## License
+
+Copyright (c) 2020 Jina AI Limited. All rights reserved.
```
```suggestion
Copyright (c) 2021 Jina AI Limited. All rights reserved.
```

nan-wang commented on multimodal-search-pdf/README.md:
```diff
@@ -0,0 +1,129 @@
+
+| Command | Description |
+| :--- | :--- |
+| ``python app.py -t index`` | To index files/data |
+| ``python app.py -t query`` | To run query on the index |
+| ``python app.py -t dryrun`` | Sanity check on the topology |
```
You have provided more options

nan-wang replied:
```
  -t, --task [index|query|query_text|query_image|query_pdf|query_restful|dryrun]
```

nan-wang replied:
`--num_docs` is not listed here

Yongxuanzhang replied:
Yes, thanks!

nan-wang commented on multimodal-search-pdf/pods/rankc.yml:
```diff
@@ -0,0 +1,16 @@
+      - jtype: Chunk2DocRankDriver
+        with:
+          traversal_paths: ['c'] #  get results for the chunks level with the results from the chunks of chunks
+      #- jtype: LogInfoDriver {}  #this can be used to format the output
```
clean up
```suggestion
```

nan-wang commented on multimodal-search-pdf/pods/index-doc.yml:
```diff
@@ -0,0 +1,32 @@
+        with:
+          fields:
+            - chunks
+      #- jtype: URI2DataURI {} # IMPORTANT: URI2DataURI can convert local uri to data uri, uncomment this to get URI data in resp
```
clean up
```suggestion
```

Yongxuanzhang replied:
Maybe we just keep it as a comment? I think this is important because if users want to have dataURI, then they can use this.

nan-wang commented on multimodal-search-pdf/pods/index-doc.yml:
```diff
@@ -0,0 +1,32 @@
+      - jtype: ExcludeQL
+        with:
+          fields:
+            - buffer
+            - chunks
+          traversal_paths: [ 'c', 'cc' ]
```
The outputs are quite clumsy right now. It's better to remove the unnecessary fields from the final results.
```suggestion
            - buffer
            - chunks
            - embedding
          traversal_paths: ['r', 'c', 'cc', 'cm', 'ccm']
```

FionnD commented on multimodal-search-pdf/README.md:
```diff
@@ -0,0 +1,129 @@
+  <img src=".github/.README_images/indexflow.png?raw=true" alt="Jina banner" width="90%">
+</p>
+
+The following image shows the structure of query flow. For query, we can send text, image and PDF files to one flow and get the results. We use 2 parallel paths to process the query data and get results.
```
two not 2

Yongxuanzhang replied:
Thanks! It's easy to forget these rules if not working on docs recently. 😂

FionnD reviewed (changes requested):
Super word! Some grammar suggestions in the README to improve it :) But it's really great, to begin with.

FionnD commented on multimodal-search-pdf/README.md:
```diff
@@ -0,0 +1,129 @@
+| Command | Description |
+| :--- | :--- |
+| ``python app.py -t index`` | To index files/data |
+| ``python app.py -t query`` | To run query on the index |
```
To run **a** query on the index

Yongxuanzhang replied:
I changed the words

FionnD commented on multimodal-search-pdf/README.md:
````diff
@@ -0,0 +1,129 @@
+When the REST gateway is enabled, Jina uses the [data URI scheme](https://en.wikipedia.org/wiki/Data_URI_scheme) to represent multimedia data. Simply organize your picture(s) into this scheme and send a POST request to `http://0.0.0.0:45670/api/search`, e.g.:
+
+``` bash
+curl --request POST -d '{"top_k": 10, "mode": "search",  "data": ["jina hello multimodal"]}' -H 'Content-Type: application/json' 'http://0.0.0.0:45670/api/search'
````
I expect users to ask us how to query other data, maybe provide an example for how to include an image in the data field? 

Could you also provide a sample result? Just the JSON output returned so user know what they expect to see back?

Yongxuanzhang replied:
Yes, this is a good suggestion! I will add the results. But it's not so easy to do curl for images or files because it's too long. 😂

Yongxuanzhang replied:
But the results are quite messy, not sure if it's ok to put here.

FionnD replied:
For files and images: can you expand sections of Markdown? I think we have feature in our readme. 

I would do it despite the bad results.

Yongxuanzhang replied:
yes

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/segmenters.py

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/segmenters.py

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/segmenters.py

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/segmenters.py

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/segmenters.py

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/segmenters.py

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
feat[mmpdf]: fix errors

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/segmenters.py

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/segmenters.py

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/segmenters.py

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
Update multimodal-search-pdf/Dockerfile

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
Update multimodal-search-pdf/README.md

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
feat[mmpdf]: fix readme

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/rankc.yml

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang committed:
Update multimodal-search-pdf/pods/index-doc.yml

Co-authored-by: Nan Wang <nan.wang@jina.ai>

Yongxuanzhang commented on multimodal-search-pdf/README.md:
```diff
@@ -0,0 +1,129 @@
+
+</p>
+
+This is an example of using [Jina](http://www.jina.ai)'s neural search framework to search text, image and PDF files in indexed PDF data.
```
Thanks a lot!!

Yongxuanzhang commented on multimodal-search-pdf/README.md:
```diff
@@ -0,0 +1,129 @@
+
+What's included in this example:
+
+- Search text, image, pdf (better partial pdf) all in one flow or in separate flows
```
I thought it might be better to search some pages of the PDF, I will remove this to avoid confusion. 😄

Yongxuanzhang committed:
feat[mmpdf]: fix readme

Yongxuanzhang commented on multimodal-search-pdf/README.md:
```diff
@@ -0,0 +1,129 @@
+
+- Search text, image, pdf (better partial pdf) all in one flow or in separate flows
+- Leverage Jina Recursive Document Representation for to segment and encode text
+- Parallel peas to reduce indexing time
```
Thanks!

Yongxuanzhang committed:
feat[mmpdf]: fix readme

Yongxuanzhang commented on multimodal-search-pdf/README.md:
```diff
@@ -0,0 +1,129 @@
+## Use toy data
+
+We have included several PDF blogs as toy data in [`toy_data`](toy_data). This data is ready to use with this example.
+You can also use other PDF files supported by [`pdfplumber`](https://github.com/jsvine/pdfplumber).
```
Yes that's great!

Yongxuanzhang commented on multimodal-search-pdf/README.md:
````diff
@@ -0,0 +1,129 @@
+## Install
+
+```bash
+pip install -r requirements.txt
````
Git clone is not included in other examples as well, I think it should be ok.

Yongxuanzhang commented on multimodal-search-pdf/README.md:
```diff
@@ -0,0 +1,129 @@
+
+## Understand the Flows
+
+The following image shows the structure of index flow, the text and image will be extracted from PDF files as chunks. Then we will use 3 paths to process and store the data.
```
Oh yes, thx!

Yongxuanzhang committed:
feat[mmpdf]: fix readme

Yongxuanzhang committed:
feat[mmpdf]: fix yaml

Yongxuanzhang committed:
feat[mmpdf]: fix readme

Yongxuanzhang committed:
feat[mmpdf]: fix readme

Yongxuanzhang committed:
feat[mmpdf]: change parallel

Yongxuanzhang committed:
feat[mmpdf]: remove unused code

Yongxuanzhang committed:
feat[mmpdf]: remove parallel

Yongxuanzhang committed:
feat[mmpdf]: add parallel for test

Yongxuanzhang committed:
feat[mmpdf]: add conftest

Yongxuanzhang committed:
Merge branch 'master' into multimodal-search-pdf

Yongxuanzhang committed:
test: add get data

Yongxuanzhang committed:
test: add summary for readme

Yongxuanzhang committed:
test: add performance metrics

Yongxuanzhang commented:
> @Yongxuanzhang if this is fully functional we should be good to merge this?
> you could add metrics around index time, evaluation, etc. here to get an idea. would be great to have a summary of the latest functional changes for this example. Remove redundant / duplicate class files.

@rutujasurve94 I added the performance in README and addressed all the major comments. I think it would be ok to go. 
Maybe later we could add a frontend for a better user experience.

rutujasurve94 commented:
LGTM 👍 Thanks for the amazing work on this - we have a strong ground to base our future examples wrt user documentation and result benchmarking due to this.

FionnD reviewed (approved):
Looks great to me :)

Yongxuanzhang commented:
Need to fix this in the latest version. :(
